components/dataHub/ioService.c

```diff
@@ -1196,6 +1196,89 @@ static void CallUpdateStartEndHandlers
 }
 
 
+//--------------------------------------------------------------------------------------------------
+/**
+ * Recursively walk the branch of the resource tree rooted at a given node, deleting anything that
+ * doesn't have admin settings or children and converting anything that does have admin settings or
+ * children into a placeholder or namespace.
+ */
+//--------------------------------------------------------------------------------------------------
+static void CleanUp
+(
+    resTree_EntryRef_t resource
+)
+//--------------------------------------------------------------------------------------------------
+{
+    // Do depth-first recursion.
+    resTree_EntryRef_t child = resTree_GetFirstChild(resource);
+    while (child != NULL)
+    {
+        resTree_EntryRef_t nextChild = resTree_GetNextSibling(child);
+        CleanUp(child);
+        child = nextChild;
+    }
+
+    switch (resTree_GetEntryType(resource))
+    {
+        case ADMIN_ENTRY_TYPE_NAMESPACE:
+        case ADMIN_ENTRY_TYPE_PLACEHOLDER:
+
+            // These don't need to be deleted.
+            // A Namespace will automatically be cleaned up when all its children go away.
+            // A Placeholder should be kept until the Admin app removes all its admin settings.
+            break;
+
+        case ADMIN_ENTRY_TYPE_INPUT:
+        case ADMIN_ENTRY_TYPE_OUTPUT:
+
+            // If this is an Input or Output resource, delete it now.
+            // This will convert it to a Placeholder if it has administrative settings.
+            resTree_DeleteIO(resource);
+            break;
+
+        case ADMIN_ENTRY_TYPE_OBSERVATION:
+        case ADMIN_ENTRY_TYPE_NONE:
+
+            // These should never be seen in an I/O API app namespace.
+            LE_FATAL("Unexpected resource type found in app's namespace.");
+            break;
```

Comment:
There is no `default` case.  Shouldn't there be an `LE_FATAL` in that case?

Reply:
A shared service such as the data hub should be resilient to bad input from clients so that one bad client doesn't affect the entire system.  So in other words, don't call `LE_FATAL` because someone passed a bad input to a public API function.  Having said that, I think if the internal state of the data hub gets messed up and internal functions start to see data that should never be possible, then it's better to abort immediately than to just ignore the problem and hope that things keep working.  In my experience, trying to keep working under buggy conditions is a recipe for *very* hard to find bugs.

Reply:
I agree with David on this one.  The lack of a default is intentional, so the compiler can catch missing cases if someone adds to the enum.  I'll add an LE_FATAL() to catch corruption of the entry type.